CMake and Visual Studio support

vii1 opened the pull request:
CMake makefile.
Some conditionals for MSVC which lacks some POSIX headers.
Provided alternatives for some GCC extensions.
Also changed SDL_WINDOW_FULLSCREEN to SDL_WINDOW_FULLSCREEN_DESKTOP, which doesn't change the screen video mode and is friendlier, at least in Windows.

tobier commented:
We really need this for easier cross-platform building.

@vii1 Can you squash your PR into a single commit and rebase on origin/master?

vii1 commented:
> @vii1 Can you squash your PR into a single commit and rebase on origin/master?
I think I got it. 👍

tobier commented:
> > @vii1 Can you squash your PR into a single commit and rebase on origin/master?
> 
> I think I got it.

LGTM but I'm not in a position of authority :+1:

mist64 commented:
Looks great. Will try out later today.

madebr commented on CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+    video.h
+)
+find_file(ROM_LABELS rom_labels.h PATHS ${CMAKE_CURRENT_SOURCE_DIR} NO_DEFAULT_PATH)
+if(NOT ROM_LABELS STREQUAL "ROM_LABELS-NOTFOUND")
```
```suggestion
if(NOT ROM_LABELS)
    list(APPEND HEADERS ${ROM_LABELS})
endif()
```
```

vii1 replied:
Why the change in the if? Did you mean `if(ROM_LABELS)`?

vii1 replied:
Ok, I see `if(ROM_LABELS)` does the same as I intended.

madebr commented on ps2.c:
```diff
@@ -72,8 +84,8 @@ ps2_step()
 			}
-
-			data_bits = current_byte << 1 | (1 - __builtin_parity(current_byte)) << 9 | (1 << 10);
+			
```
whitespace

madebr commented on ps2.c:
```diff
@@ -47,6 +47,18 @@ static int send_state = 0;
 int ps2_clk_out, ps2_data_out;
 int ps2_clk_in, ps2_data_in;
 
+#ifndef __GNUC__
```
- __popcnt16, __popcnt or __popcnt64 for Visual Studio?
https://docs.microsoft.com/en-us/cpp/intrinsics/popcnt16-popcnt-popcnt64?view=vs-2019
- Keep it a macro and use uppercase?

vii1 replied:
Not sure about this… docs say cpuid must be checked at runtime to ensure popcnt is supported.

mist64 replied:
Please only keep the generic version. There is no point in optimizing this per compiler and making the code more complicated and more fragile. Besides, I bet most compilers will detect this pattern and generate the same code as `__builtin_parity(x)` anyway.

(Also: tabs, not spaces)

vii1 replied:
Done!

madebr commented on cmake/FindSDL2.cmake:
```diff
@@ -0,0 +1,173 @@
+
+INCLUDE(FindPackageHandleStandardArgs)
+
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(SDL2 REQUIRED_VARS SDL2_LIBRARY SDL2_INCLUDE_DIR)
```
I modified `FindSDL2.cmake` in a project to add a `SDL2::SDL2` target:
https://github.com/rwengine/openrw/blob/51b7264744d1aaa20de3b86a7a4e92a9930881ba/cmake/modules/FindSDL2.cmake#L210-L216

This alllows usage of SDL2 as:
target_link_libraries(x16emu PUBLIC SDL2::SDL2)

vii1 replied:
Neat! Done.

madebr commented on CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+    set(HEADERS ${HEADERS} ${ROM_LABELS})
+endif()
+add_executable(x16emu ${SOURCES} ${HEADERS})
+set(LIBS ${SDL2_LIBRARY} ${SDL2MAIN_LIBRARY})
```
The `SDL2main` library` is not available in all linux distros when having installed `SDL2-devel`.
I have fixed this in my fork: https://github.com/madebr/x16-emulator/commits/pr64
See https://wiki.libsdl.org/SDL_SetMainReady

vii1 replied:
Removed the dependency. I think it's only needed if you want a pure GUI app in Windows, maybe in other systems too.

madebr replied:
I'll rebase my changes on your pr. Right now, I'm working on adding emscripten support such that the cmake and make builds are equivalent.

madebr commented on debugger.h:
```diff
@@ -12,7 +12,7 @@
 #define _DEBUGGER_H
 
-#include <SDL.h>
+#include <SDL2/SDL.h>
```
This change contradicts with the remark in the `FindSDL2.cmake` script:
```
# Note that the header path has changed from SDL2/SDL.h to just SDL.h
# This needed to change because "proper" SDL convention
# is #include "SDL.h", not <SDL2/SDL.h>. This is done for portability
# reasons because not all systems place things in SDL2/ (see FreeBSD).
```

vii1 replied:
You're right. This is also noted [in the FAQ](https://wiki.libsdl.org/FAQDevelopment#Do_I_.23include_.3CSDL.h.3E_or_.3CSDL.2FSDL.h.3E.3F), so I'll change the includes to follow that.

madebr commented on loadsave.c:
```diff
@@ -7,7 +7,9 @@
 #include <stdbool.h>
 #include <sys/stat.h>
 #include <dirent.h>
```
This file does not exist on Visual Studio (2019).

vii1 replied:
I added this alternative https://github.com/tronkko/dirent
I had it installed globally via vcpkg. Thanks for noting.

mist64 commented:
I'll keep following your conversation for a little while longer. :)

madebr commented:
@mist64 I've got appveyor CI working (including artifacts)!
You should let the admins register an organization account on appveyor.com such that https://ci.appveyor.com/project/commanderx16/x16-emulator will point to the builds.

See https://ci.appveyor.com/project/madebr/x16-emulator/builds/27537615

tobier commented:
We merged a patch yesterday about Travis CI integration for Linux and macOS, it would be great to add Windows support via AppVeyor (I believe AppVeyor only support Linux and Windows).

@vii1 @madebr if you're working on CI for CMake can you please also update the .travis.yml file I added in commit 44ec5dd. We should probably also drop the existing Makefile if we decide to go with CMake

madebr commented:
@madebr travis CI using CMake is already working: https://travis-ci.com/madebr/x16-emulator/builds/128437480

I'm currently working on adding CMake support to x16-rom, and then x16-docs.
That way,  a release can be built easily.

tobier commented:
@madebr :+1:

madebr commented:
@mist64 While working on adding emscripten support, I needed the `chargen.bin` file.
`README.md` says it is in the binary releases.
I could find it in `x16emu_linux-r30.zip` but not in `x16emu_linux-r31.zip`.
Also, wouldn't it be easier to add it to the rom repo? (Including steps how to generate it)

vii1 commented:
> 
> 
> @mist64 While working on adding emscripten support, I needed the `chargen.bin` file.
> `README.md` says it is in the binary releases.
> I could find it in `x16emu_linux-r30.zip` but not in `x16emu_linux-r31.zip`.
> Also, wouldn't it be easier to add it to the rom repo? (Including steps how to generate it)

Chargen data is now included in `rom.bin`, so `chargen.bin` is not necessary anymore.

madebr commented on CMakeLists.txt:
```diff
@@ -3,44 +3,45 @@ cmake_minimum_required(VERSION 3.0)
+	target_link_libraries(x16emu m)
 endif()
-target_link_libraries(x16emu ${LIBS})
+target_link_libraries(x16emu PUBLIC SDL2::SDL2)
```
No need to make it public
```suggestion
target_link_libraries(x16emu PRIVATE SDL2::SDL2)
```

madebr commented on CMakeLists.txt:
```diff
@@ -3,44 +3,45 @@ cmake_minimum_required(VERSION 3.0)
-set(LIBS ${SDL2_LIBRARY} ${SDL2MAIN_LIBRARY})
 if(NOT MSVC)
-    set(LIBS ${LIBS} m)
+	target_link_libraries(x16emu m)
```
If you use INTERFACE, PRIVATE or PUBLIC once on a target, CMake requires consistent use.
```suggestion
	target_link_libraries(x16emu PRIVATE m)
```

madebr commented on CMakeLists.txt:
```diff
@@ -3,44 +3,45 @@ cmake_minimum_required(VERSION 3.0)
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_LIST_DIR}/cmake")
 find_package(SDL2 REQUIRED)
-include_directories(${SDL2_INCLUDE_DIR})
+if(WIN32)
```
mingw has `dirent.h`(https://github.com/Alexpux/mingw-w64/blob/master/mingw-w64-headers/crt/dirent.h)
```suggestion
if(MSVC)
```

madebr commented on CMakeLists.txt:
```diff
@@ -3,44 +3,45 @@ cmake_minimum_required(VERSION 3.0)
 find_package(SDL2 REQUIRED)
-include_directories(${SDL2_INCLUDE_DIR})
+if(WIN32)
+	include_directories("${CMAKE_CURRENT_LIST_DIR}/windows")
```
I would put it in another folder.
Maybe in `msvc/include` or `support/msvc/include`.
Maybe @mist64 has a preference

vii1 replied:
I moved it to `support/msvc/include` for now.

madebr commented on debugger.c:
```diff
@@ -12,7 +12,7 @@
 #include <stdio.h>
 #include <inttypes.h>
-#include <SDL2/SDL.h>
+#include "SDL.h"
```
```suggestion
#include <SDL.h>
```
(and elsewhere)

greg-king5 replied:
It isn't a "system header"; it's a "user header".  That usage isn't about where the file sits in the file-system.  It's about how the preprocessor knows where to find it.  System directories already are known to the preprocessor.  But, "/usr/include/SDL2/" isn't known; the command line uses a `-I` option to tell the preprocessor where to look.  That makes it "a user-specified path".

madebr replied:
That distinction is new to me.
By my knowledge, user headers are files of the current project, other wise system.
I thought that's why we have `-I` and `-isystem`.
Can you point me to a text giving reasons for this because I'm not convinced :)

vii1 replied:
I chose quotes as the [SDL FAQ](https://wiki.libsdl.org/FAQDevelopment#Do_I_.23include_.3CSDL.h.3E_or_.3CSDL.2FSDL.h.3E.3F) advises:

> ### Do I #include <SDL.h> or <SDL/SDL.h>?
> 
> The most portable way to include SDL headers is to use quotes around the header name:
> 
> #include "SDL.h"

madebr replied:
Can't argue against portability :D

madebr commented on debugger.h:
```diff
@@ -12,7 +12,7 @@
 #define _DEBUGGER_H
 
-#include <SDL2/SDL.h>
+#include "SDL.h"
```
```suggestion
#include <SDL.h>
```

madebr commented on ps2.c:
```diff
@@ -47,17 +47,14 @@ static int send_state = 0;
 
-#ifndef __GNUC__
-int parity( uint8_t b )
+int
```
```suggestion
static int
```

madebr commented on video.h:
```diff
@@ -7,7 +7,8 @@
 #include <stdbool.h>
 #include <stdint.h>
-#include <SDL2/SDL.h>
+#define SDL_MAIN_HANDLED
```
I would put this define at the top of the `main.c` file (or wherever the `main` function is defined).
Otherwise reordering of includes can introduce a bug.

vii1 replied:
Ooops. I meant to put it in `video.c`. I put it there instead of `main.c` because there's no reference to SDL in `main.c`. SDL initialization is made in `video.c`.

vii1 replied:
Never mind, I was wrong. `main.c` includes `video.h` which includes `SDL.h`.

madebr commented on CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+project(x16emu)
+cmake_minimum_required(VERSION 3.0)
+
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_LIST_DIR}/cmake")
```
```suggestion
list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")
```

madebr commented on main.c:
```diff
@@ -2,7 +2,8 @@
 
-#define _XOPEN_SOURCE   600
+#define SDL_MAIN_HANDLED
+#define _XOPEN_SOURCE 600
```
Why is this required?

vii1 replied:
If you want a pure GUI app in Windows (i.e. with no console), you need to define a `WinMain` function. SDL offers the posibility of using just the same `main` for all platforms, by `#define`ing `main` as `SDL_main` and providing a stub `main` and `WinMain` in `SDLmain.lib`. It happens something similar in macOS, but there you compile `SDLmain.m` (which provides an Objective-C entrypoint) into your project.

Since x16emu does not need to be a pure GUI program and uses the console to provide logs and messages and so, when you said some *nix distros do not provide `SDLmain.a`, I thought it appropiate to just get rid of that dependence. But when you do that, you need to define `SDL_MAIN_HANDLED` before including SDL so the SDL headers do not `#define main`.

greg-king5 replied:
I think that he asked about the `_XOPEN_SOURCE` macro.

vii1 replied:
Oh. No idea about that.

madebr replied:
Nevermind my remark.
I did not notice that you modified the number of spaces in that define.

madebr commented on CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+project(x16emu)
+cmake_minimum_required(VERSION 3.0)
```
These 2 lines should be switched.
First call `cmake_minimum_required`, then `project`.

madebr replied:
Also, I would do
```suggestion
project(x16emu LANGUAGES C)
```
since CXX is not used.

madebr commented:
This all looks good to me.
Once this gets merged, I'll open a new PR that will add all functionality of the current Makefile.

mist64 commented:
I have been watching this from a distance from a while. I am not an expert on CMake, but I agree that it is a good build system, and I'm not opposed to switching to it, especially if it gives us Visual Studio support.

So what is the status of this? Is it okay to merge it, will it be able to do the existing builds I do with make package_mac/win/linux? (Windows is currently cross-compiled for mingw from a Mac.)

madebr commented:
Hey, I have tested this pr and it will build fine on linux/mac/visual studio.
(I have not tested mingw on Windows)
It has some missing bits:
- For mingw on Linux, it needs a small patch (at least on my Fedora 30 machine)
- no (optional) TM2151 support
- no documentation
- no emscripten/webassembly support

Because I did not want to interfere with @vii1 's pr, I will open a pr once this gets merged.
This pr will also add ci for linux and windows.

vii1 commented:
I tested it in Visual Studio 19 and mingw under WSL (Ubuntu) in Windows. For mingw I used [this toolchain scripts](https://github.com/mosra/toolchains/tree/master/archlinux)

madebr commented:
@vii1 Can you test [this mingw toolchain file](https://github.com/madebr/x16-emulator/blob/cmake/cmake/toolchains/mingw.cmake) which I intend to add in a next pr?
It should be distribution independent as no paths are explicitly set.
You must set the `CC` environment variable or `CMAKE_C_COMPILER` cmake variable manually to either a 32 or 64 bit mingw C compiler.

vii1 commented:
@madebr It didn't detect my SDL2 location. I put it explicitly but build failed. Tried both setting `CC` and `-DCMAKE_C_COMPILER`. It seems it doesn't link the required Windows libraries.

[See log](https://github.com/commanderx16/x16-emulator/files/3659063/log.txt)

`x86_64-w64-mingw32-gcc -print-sysroot` yields nothing.

madebr commented:
@vii1 Looking at your output, there is no C compiler detection.
<details><summary>Output of running CMake with mingw on Fedora</summary>
<p>

```bash
$ CC=x86_64-w64-mingw32-gcc cmake .. -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/mingw.cmake
-- The C compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/bin/x86_64-w64-mingw32-gcc
-- Check for working C compiler: /usr/bin/x86_64-w64-mingw32-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Development build
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Found SDL2: mingw32;-mwindows;/usr/x86_64-w64-mingw32/sys-root/mingw/lib/libSDL2main.a;/usr/x86_64-w64-mingw32/sys-root/mingw/lib/libSDL2.dll.a  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/maarten/programming/commander-x16/x16-emulator/build
```
</p>
</details>

vii1 commented:
@madebr Maybe this can help https://github.com/alainfrisch/flexdll/commit/3a7fe095e6c723c3054363a4fe44de44cf480aa4

madebr commented:
@vii1 I modified the mingw toolchain script based upon your link.
Can you clean the build dir and try again?

vii1 commented:
It detected everything except the SDL2 include directory. I set it explicitly and then it worked.

[log.txt](https://github.com/commanderx16/x16-emulator/files/3664275/log.txt)

madebr commented:
@vii1 Weird. It's detecting the library, but not the include directory. Which means cmake knows about the root directory.
To debug:
- This is Arch Linux under WSL, correct?
- What is the output of `x86_64-w64-mingw32-gcc -print-search-dirs`?
- Can you rerun the mingw toolchain of my fork and copy paste the line about `CMAKE_FIND_ROOT_PATH`?
- SDL2 include directory is at `/usr/x86_64-w64-mingw32/include/SDL2`?
- SDL2.dll is at `/usr/x86_64-w64-mingw32/bin/SDL2.dll`?
- Isn't there a SDL2.dll.a (or similar static library)?

MouriNaruto commented:
Good job, and I also think this project should use CMake instead of Makefile.

madebr commented:
@vii1 I just tried out archlinux/base in a docker and could compile and link x16emu using the [mingw-w64 packages of ownstuff](https://wiki.archlinux.org/index.php/Unofficial_user_repositories#ownstuff)
So I think the toolchain is fine.
For reference:
```
x86_64-w64-mingw32-gcc -print-search-dirs
install: /usr/lib/gcc/x86_64-w64-mingw32/9.2.0/
programs: =/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/:/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/:/usr/lib/gcc/x86_64-w64-mingw32/:/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/:/usr/lib/gcc/x86_64-w64-mingw32/:/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/x86_64-w64-mingw32/9.2.0/:/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/
libraries: =/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/:/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/lib/x86_64-w64-mingw32/9.2.0/:/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/lib/../lib/:/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/lib/
```
More importantly:
Because #127 moved some files, you'll need to fix the CMake file:
- add include directories to `extern/src` and `extern/include` (`include_directories("${CMAKE_CURRENT_LIST_DIR}/extern/include" "${CMAKE_CURRENT_LIST_DIR}/extern/src"`)
- Maybe move `support/msvc/include/dirent.h` to `extern/include/msvc/dirent.h` and include that directory?

vii1 commented:
@madebr
> * This is Arch Linux under WSL, correct?

No. This is Ubuntu 18.04 LTS under WSL. I just used an Arch toolchain file but all paths match. Sorry about the confusion.

> * What is the output of `x86_64-w64-mingw32-gcc -print-search-dirs`?

```
install: /usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/
programs: =/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/:/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/:/usr/lib/gcc/x86_64-w64-mingw32/:/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/:/usr/lib/gcc/x86_64-w64-mingw32/:/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/../../../../x86_64-w64-mingw32/bin/x86_64-w64-mingw32/7.3-win32/:/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/../../../../x86_64-w64-mingw32/bin/
libraries: =/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/:/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/../../../../x86_64-w64-mingw32/lib/x86_64-w64-mingw32/7.3-win32/:/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/../../../../x86_64-w64-mingw32/lib/
```
(how do you insert those neat disclosable blocks?)

> * Can you rerun the mingw toolchain of my fork and copy paste the line about `CMAKE_FIND_ROOT_PATH`? (I've added a debug statement)

```
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/7.3-win32/../../../../x86_64-w64-mingw32/lib//..
```
Both paths are correct. That weird long path resolves to `/usr/x86_64-w64-mingw32`.

> * SDL2 include directory is at `/usr/x86_64-w64-mingw32/include/SDL2`?

Yes.

> * SDL2.dll is at `/usr/x86_64-w64-mingw32/bin/SDL2.dll`?

Yes.

> * Isn't there a SDL2.dll.a (or similar static library)?

Yes. This is what `cd /usr/x86_64-w64-mingw32/lib; ls *SDL*` yields:
```
libSDL2.a
libSDL2.dll.a
libSDL2.la
libSDL2_test.a
libSDL2_test.la
libSDL2main.a
libSDL2main.la
```

madebr commented:
@vii1 I just tried the Ubuntu docker and got the same result.
Because apt does not provide mingw packages, I extracted the mingw runtime directories of http://libsdl.org/download-2.0.php to the root.

<details><summary>Ubuntu Linux mingw-w64 output</summary>
<p>

```bash
root@0421053809e7:/x16/build# cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=19.04
DISTRIB_CODENAME=disco
DISTRIB_DESCRIPTION="Ubuntu 19.04"
root@0421053809e7:/x16/build# CC=x86_64-w64-mingw32-gcc cmake .. -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/mingw.cmake
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
-- The C compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/bin/x86_64-w64-mingw32-gcc
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
-- Check for working C compiler: /usr/bin/x86_64-w64-mingw32-gcc -- works
-- Detecting C compiler ABI info
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
-- Detecting C compiler ABI info - done
-- Detecting C compile features
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
-- Detecting C compile features - done
-- Development build
-- Looking for pthread.h
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
-- Looking for pthread.h - found
-- Looking for pthread_create
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
Setting CMAKE_FIND_ROOT_PATH to /usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib//..
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found SDL2: mingw32;-mwindows;/usr/lib/gcc/x86_64-w64-mingw32/8.3-win32/../../../../x86_64-w64-mingw32/lib/../bin/SDL2.dll;-lpthread  
-- Configuring done
-- Generating done
-- Build files have been written to: /x16/build
root@0421053809e7:/x16/build# make VERBOSE=y                                                                         
/usr/bin/cmake -S/x16 -B/x16/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /x16/build/CMakeFiles /x16/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/x16/build'
make -f CMakeFiles/x16emu.dir/build.make CMakeFiles/x16emu.dir/depend
make[2]: Entering directory '/x16/build'
cd /x16/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /x16 /x16 /x16/build /x16/build /x16/build/CMakeFiles/x16emu.dir/DependInfo.cmake --color=
Dependee "/x16/build/CMakeFiles/x16emu.dir/DependInfo.cmake" is newer than depender "/x16/build/CMakeFiles/x16emu.dir/depend.internal".
Dependee "/x16/build/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/x16/build/CMakeFiles/x16emu.dir/depend.internal".
Scanning dependencies of target x16emu
make[2]: Leaving directory '/x16/build'
make -f CMakeFiles/x16emu.dir/build.make CMakeFiles/x16emu.dir/build
make[2]: Entering directory '/x16/build'
[  7%] Building C object CMakeFiles/x16emu.dir/cpu/fake6502.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/cpu/fake6502.c.obj   -c /x16/cpu/fake6502.c
[ 15%] Building C object CMakeFiles/x16emu.dir/debugger.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/debugger.c.obj   -c /x16/debugger.c
[ 23%] Building C object CMakeFiles/x16emu.dir/disasm.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/disasm.c.obj   -c /x16/disasm.c
[ 30%] Building C object CMakeFiles/x16emu.dir/loadsave.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/loadsave.c.obj   -c /x16/loadsave.c
[ 38%] Building C object CMakeFiles/x16emu.dir/main.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/main.c.obj   -c /x16/main.c
[ 46%] Building C object CMakeFiles/x16emu.dir/memory.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/memory.c.obj   -c /x16/memory.c
[ 53%] Building C object CMakeFiles/x16emu.dir/ps2.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/ps2.c.obj   -c /x16/ps2.c
[ 61%] Building C object CMakeFiles/x16emu.dir/sdcard.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/sdcard.c.obj   -c /x16/sdcard.c
[ 69%] Building C object CMakeFiles/x16emu.dir/spi.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/spi.c.obj   -c /x16/spi.c
[ 76%] Building C object CMakeFiles/x16emu.dir/vera_spi.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/vera_spi.c.obj   -c /x16/vera_spi.c
[ 84%] Building C object CMakeFiles/x16emu.dir/via.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/via.c.obj   -c /x16/via.c
[ 92%] Building C object CMakeFiles/x16emu.dir/video.c.obj
/usr/bin/x86_64-w64-mingw32-gcc  @CMakeFiles/x16emu.dir/includes_C.rsp -O3 -Wall -Werror -std=c99 -o CMakeFiles/x16emu.dir/video.c.obj   -c /x16/video.c
[100%] Linking C executable x16emu.exe
/usr/bin/cmake -E cmake_link_script CMakeFiles/x16emu.dir/link.txt --verbose=y
/usr/bin/cmake -E remove -f CMakeFiles/x16emu.dir/objects.a
/usr/bin/x86_64-w64-mingw32-ar cr CMakeFiles/x16emu.dir/objects.a @CMakeFiles/x16emu.dir/objects1.rsp
/usr/bin/x86_64-w64-mingw32-gcc   -Wl,--subsystem,console -Wl,--whole-archive CMakeFiles/x16emu.dir/objects.a -Wl,--no-whole-archive  -o x16emu.exe -Wl,--out-implib,libx16emu.dll.a -Wl,--major-image-version,0,--minor-image-version,0 @CMakeFiles/x16emu.dir/linklibs.rsp
make[2]: Leaving directory '/x16/build'
[100%] Built target x16emu
make[1]: Leaving directory '/x16/build'
/usr/bin/cmake -E cmake_progress_start /x16/build/CMakeFiles 0
```
</p>
</details>

<details>
<summary>sdl files in the mingw tree</summary>
<p>

```bash
root@0421053809e7:/x16/build# find /usr/x86_64-w64-mingw32/ -iname "*sdl*"
/usr/x86_64-w64-mingw32/bin/sdl2-config
/usr/x86_64-w64-mingw32/bin/SDL2.dll
/usr/x86_64-w64-mingw32/include/SDL2
/usr/x86_64-w64-mingw32/include/SDL2/SDL_pixels.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_revision.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_platform.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_blendmode.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_scancode.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_loadso.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_clipboard.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_audio.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_video.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_syswm.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_types.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_keyboard.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_log.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_assert.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_rwops.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_sensor.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_rect.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_cpuinfo.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_gesture.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_opengl_glext.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_endian.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_quit.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_haptic.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_bits.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_gamecontroller.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_assert.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_md5.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_harness.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_error.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_main.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2_gl2.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_common.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_name.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_shape.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_keycode.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_joystick.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_crc32.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_power.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_config.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_version.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_messagebox.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_font.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_random.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_thread.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_images.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_fuzzer.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_events.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_touch.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_surface.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2_gl2ext.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_system.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_compare.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_mouse.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_opengl.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_filesystem.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_timer.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_hints.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2_khrplatform.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_egl.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_test_memory.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_vulkan.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2_gl2platform.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_mutex.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_atomic.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_log.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_stdinc.h
/usr/x86_64-w64-mingw32/include/SDL2/SDL_render.h
/usr/x86_64-w64-mingw32/include/rasdlg.h
/usr/x86_64-w64-mingw32/lib/libSDL2.dll.a
/usr/x86_64-w64-mingw32/lib/libSDL2_test.a
/usr/x86_64-w64-mingw32/lib/pkgconfig/sdl2.pc
/usr/x86_64-w64-mingw32/lib/libSDL2_test.la
/usr/x86_64-w64-mingw32/lib/cmake/SDL2
/usr/x86_64-w64-mingw32/lib/cmake/SDL2/sdl2-config.cmake
/usr/x86_64-w64-mingw32/lib/libSDL2.la
/usr/x86_64-w64-mingw32/lib/librasdlg.a
/usr/x86_64-w64-mingw32/lib/libSDL2main.la
/usr/x86_64-w64-mingw32/lib/libSDL2main.a
/usr/x86_64-w64-mingw32/lib/libSDL2.a
```
</p>
</details>

As you can see, our build trees are completely equivalent.

vii1 commented:
> Because #127 moved some files, you'll need to fix the CMake file:
> 
> * add include directories to `extern/src` and `extern/include` (`include_directories("${CMAKE_CURRENT_LIST_DIR}/extern/include" "${CMAKE_CURRENT_LIST_DIR}/extern/src"`)
> * Maybe move `support/msvc/include/dirent.h` to `extern/include/msvc/dirent.h` and include that directory?

@madebr Merged and done. I also added support for `-DWITH_YM2151`.

> @vii1 I just tried the Ubuntu docker and got the same result.
> Because apt does not provide mingw packages, I extracted the mingw runtime directories of http://libsdl.org/download-2.0.php to the root.
> 
> Ubuntu Linux mingw-w64 output
> sdl files in the mingw tree
> As you can see, our build trees are completely equivalent.

Then probably there's a mistake at my side. I think we can move over.

madebr commented:
@vii1 
I just tried it myself on WSL and could briefly reproduce a similar issue as you have.
The problem I had was that the copied SDL2 files in the `/usr/x86_64-w64-mingw32` directory did not have the correct file mode bits and I did not copy the files correctly. (This is the first time I'm using WSL)

Try running this command and compare your output:
```bash
find /usr/x86_64-w64-mingw32/ -iname "*sdl*"  | (while read -r LINE;  do ls -dal "$LINE"; done;)
```
<details>
<summary>Output of command above on my WSL (working mingw build)</summary>

```bash
$ find /usr/x86_64-w64-mingw32/ -iname "*sdl*"  | (while read -r LINE;  do ls -dal "$LINE"; done;)
-rwxr-xr-x 1 root root 1401344 Sep 29 15:24 /usr/x86_64-w64-mingw32/bin/SDL2.dll
-rwxr-xr-x 1 root root 1444 Sep 29 15:24 /usr/x86_64-w64-mingw32/bin/sdl2-config
drwxr-xr-x 1 root root 512 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2
-rw-r--r-- 1 root root 4210 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL.h
-rw-r--r-- 1 root root 11045 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_assert.h
-rw-r--r-- 1 root root 10839 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_atomic.h
-rw-r--r-- 1 root root 35788 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_audio.h
-rw-r--r-- 1 root root 3096 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_bits.h
-rw-r--r-- 1 root root 5115 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_blendmode.h
-rw-r--r-- 1 root root 1966 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_clipboard.h
-rw-r--r-- 1 root root 6776 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_config.h
-rw-r--r-- 1 root root 8610 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_cpuinfo.h
-rw-r--r-- 1 root root 73586 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_egl.h
-rw-r--r-- 1 root root 6451 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_endian.h
-rw-r--r-- 1 root root 2271 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_error.h
-rw-r--r-- 1 root root 30883 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_events.h
-rw-r--r-- 1 root root 5255 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_filesystem.h
-rw-r--r-- 1 root root 13531 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_gamecontroller.h
-rw-r--r-- 1 root root 2157 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_gesture.h
-rw-r--r-- 1 root root 39294 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_haptic.h
-rw-r--r-- 1 root root 53343 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_hints.h
-rw-r--r-- 1 root root 13836 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_joystick.h
-rw-r--r-- 1 root root 6437 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_keyboard.h
-rw-r--r-- 1 root root 15262 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_keycode.h
-rw-r--r-- 1 root root 2866 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_loadso.h
-rw-r--r-- 1 root root 6491 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_log.h
-rw-r--r-- 1 root root 5375 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_main.h
-rw-r--r-- 1 root root 4611 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_messagebox.h
-rw-r--r-- 1 root root 10924 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_mouse.h
-rw-r--r-- 1 root root 6665 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_mutex.h
-rw-r--r-- 1 root root 1155 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_name.h
-rw-r--r-- 1 root root 82593 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_opengl.h
-rw-r--r-- 1 root root 731497 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_opengl_glext.h
-rw-r--r-- 1 root root 1254 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles.h
-rw-r--r-- 1 root root 1552 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2.h
-rw-r--r-- 1 root root 31876 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2_gl2.h
-rw-r--r-- 1 root root 98695 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2_gl2ext.h
-rw-r--r-- 1 root root 913 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2_gl2platform.h
-rw-r--r-- 1 root root 10022 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_opengles2_khrplatform.h
-rw-r--r-- 1 root root 17188 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_pixels.h
-rw-r--r-- 1 root root 5203 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_platform.h
-rw-r--r-- 1 root root 2463 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_power.h
-rw-r--r-- 1 root root 2106 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_quit.h
-rw-r--r-- 1 root root 4839 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_rect.h
-rw-r--r-- 1 root root 42721 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_render.h
-rw-r--r-- 1 root root 79 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_revision.h
-rw-r--r-- 1 root root 9197 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_rwops.h
-rw-r--r-- 1 root root 15254 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_scancode.h
-rw-r--r-- 1 root root 7545 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_sensor.h
-rw-r--r-- 1 root root 5681 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_shape.h
-rw-r--r-- 1 root root 22163 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_stdinc.h
-rw-r--r-- 1 root root 20632 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_surface.h
-rw-r--r-- 1 root root 9152 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_system.h
-rw-r--r-- 1 root root 9252 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_syswm.h
-rw-r--r-- 1 root root 2000 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test.h
-rw-r--r-- 1 root root 3243 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_assert.h
-rw-r--r-- 1 root root 5732 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_common.h
-rw-r--r-- 1 root root 2163 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_compare.h
-rw-r--r-- 1 root root 3385 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_crc32.h
-rw-r--r-- 1 root root 2439 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_font.h
-rw-r--r-- 1 root root 13122 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_fuzzer.h
-rw-r--r-- 1 root root 4612 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_harness.h
-rw-r--r-- 1 root root 2215 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_images.h
-rw-r--r-- 1 root root 1954 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_log.h
-rw-r--r-- 1 root root 4630 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_md5.h
-rw-r--r-- 1 root root 1787 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_memory.h
-rw-r--r-- 1 root root 3156 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_test_random.h
-rw-r--r-- 1 root root 12977 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_thread.h
-rw-r--r-- 1 root root 3454 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_timer.h
-rw-r--r-- 1 root root 2946 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_touch.h
-rw-r--r-- 1 root root 1031 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_types.h
-rw-r--r-- 1 root root 5162 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_version.h
-rw-r--r-- 1 root root 46512 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_video.h
-rw-r--r-- 1 root root 10797 Sep 29 15:54 /usr/x86_64-w64-mingw32/include/SDL2/SDL_vulkan.h
lrwxrwxrwx 1 root root 38 Nov 18  2017 /usr/x86_64-w64-mingw32/include/rasdlg.h -> ../../share/mingw-w64/include/rasdlg.h
drwxr-xr-x 1 root root 512 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/cmake/SDL2
-rw-r--r-- 1 root root 482 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/cmake/SDL2/sdl2-config.cmake
-rw-r--r-- 1 root root 13687676 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/libSDL2.a
-rwxr-xr-x 1 root root 427788 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/libSDL2.dll.a
-rwxr-xr-x 1 root root 1058 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/libSDL2.la
-rw-r--r-- 1 root root 537396 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/libSDL2_test.a
-rwxr-xr-x 1 root root 925 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/libSDL2_test.la
-rw-r--r-- 1 root root 18782 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/libSDL2main.a
-rwxr-xr-x 1 root root 922 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/libSDL2main.la
-rw-r--r-- 1 root root 30394 Nov 18  2017 /usr/x86_64-w64-mingw32/lib/librasdlg.a
-rw-r--r-- 1 root root 761 Sep 29 15:24 /usr/x86_64-w64-mingw32/lib/pkgconfig/sdl2.pc
```
</details>

madebr commented:
@vii1 @mist64 PR is looking good!

vii1 commented:
Merged and solved conflicts.
Removed Elvis `?:` operator which does not exist in MSVC.

y2k4life commented:
@vii1 Great work. Tested with Visual Studio 2019 no problems. The only drawback is not having the ROMs. I still had to compile in Linux due the conversion for the charset. Will you be working on building ROMs with Windows as well?

manuvdp commented:
I can't get it working on windows 64bit machine
receives this errors:
LNK2019	unresolved external symbol _SDL_RenderPresent referenced in function _video_update	D:\Temp\x16-VS\CMakeLists.txt	D:\Temp\x16-VS\video.c.obj	1

Can you please add the cMakeSettings.json, or other missing files like dirent.h.
I'm complete noob in c++ and i'm trying to learn. Can you help me with a set by step how I can get the build working? Doen it need to be build in 64bit? os x86 is ok?

In which foler should I put the /SDL/include and /SDL/lib? Just set a setting in the cMakeSettings.json?


`{
  "configurations": [
    {
      "name": "x86-Debug",
      "generator": "Ninja",
      "configurationType": "Debug",
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "",
      "buildCommandArgs": "-v",
      "ctestCommandArgs": "",
      "inheritEnvironments": [ "msvc_x86" ],
      "variables": [
        {
          "name": "SDL2_DIR",
          "value": "D:/Temp/x16-VS/include",
          "type": "PATH"
        },
        {
          "name": "SDL2_INCLUDE_DIR",
          "value": "D:/Temp/x16-VS/include",
          "type": "PATH"
        },
        {
          "name": "SDL2_LIBRARY_TEMP",
          "value": "C:\\Program Files\\SDL\\temp",
          "type": "FILEPATH"
        },
        {
          "name": "SDL2MAIN_LIBRARY",
          "value": "D:/Temp/x16-VS/lib",
          "type": "FILEPATH"
        }
      ],
      "cmakeToolchain": "D:/Temp/x16-VS/cmake/FindSDL2.cmake"
    
    }
  ]
}`

madebr commented:
@manuvdp 
The project can be built as either a 32- or a 64-bit executable. Don't forget that a 32-bit build needs to link to a 32-bit SDL2 library, and vice versa for 64-bit.

Assuming you downloaded [SDL2-devel-2.0.10-VC.zip](https://www.libsdl.org/release/SDL2-devel-2.0.10-VC.zip), you should only need to set `SDL2_PATH` to the location where you unzipped the archive.
I can't make up from your comment where you downloaded this archive. `C:\Program Files\SDL`?
You can (and should) remove the other variables as CMake will autodetect these.
Don't forget to reconfigure because CMake might remember these old variables. 

I think adding a `CMakeSettings.json` file is not useful, because the content of the `SDL2_PATH` variable will be different on every machine.

It seems `dirent.h` is currently not added to this pr, so Visual Studio will currently not be able to compile it unless [you install it using vcpkg](https://github.com/commanderx16/x16-emulator/pull/64#discussion_r326691113)

manuvdp commented:
@madebr , thanks for the fast response. These where the steps that I had to do. Perhaps good to add a readme_win.md file ?
download cmake-gui	https://cmake.org/download/
install cmake-gui
use vs 2019 x86 in Gui
ROM_LABELS 		Have no idea? Generated an empty rom_labels.h file and pointed to him
SDL2MAIN_LIBRARY 	C:/Program Files/SDL/lib/x86/SDL2main.lib
SDL2_INCLUDE_DIR 	C:/Program Files/SDL/include
SLD2_LIBRARY		C:/Program Files/SDL/lib/x86/SDL2.lib
Press 'Generate'
Press 'Open Project'

Then copy dlls from the executable x16 emulator:
- libgcc_s_sjlj-1.dll
- libwinpthread-1.dll
- SDL2.lib
- rom.bin
- rom.sym
to $$builddirectory$$/debug/

Then right click on thex16emu project and go to
-> Configuration Properties -> Debugging -> Working Directory set .\Debug

Launch


Now it runs but I have vertical lines on the screen:
![image](https://user-images.githubusercontent.com/2649830/67518221-9a947200-f6a4-11e9-9d66-548f636d27d2.png)

The debugger seems to work:
![image](https://user-images.githubusercontent.com/2649830/67518249-ae3fd880-f6a4-11e9-85e7-be852f0045c5.png)


Any idea?

madebr commented:
- `cmake-gui` is indeed useful to configure cmake, modify variables etc. Please remember that the downloaded cmake version can differ from the cmake version supplied with Visual Studio.
- Ideally, cmake should have detected these `SDL_*` variables automatically (from `SDL2_PATH`) but you can always define them yourself. Documenting this process somewhere may be useful. Adding it to a section in build_instructions.md would be useful.
- From the files you listed, only `rom.bin` must be copied to the build directory.
  * `libgcc_s_sjlj-1.dll` and `libwinpthread-1.dll` are the libc and pthread libraries of mingw. Since you are compiling and linking using Visual Studio, these are not needed.
  * `SDL2.lib`: This is a link libraries. This is only needed at link time. Not at run time.
  * `rom.bin`: This is needed. This contains the actual rom
  * `rom.sym`: This file contains the address locations of the symbols. This is useful for debugging and playing with poking at addresses.

- I don't know exactly, but I think your artifacts can be explained by this branch being out of date. The rom has been updated in the mean time. Yesterday, I have rebased my cmake branch on the master.
I just tested it on and I see no visual artifacts.
My code is at https://github.com/madebr/x16-emulator/tree/cmake_rebase
Windows CI is at https://ci.appveyor.com/project/madebr/x16-emulator/history
Note that I have uploaded artifacts: e.g. https://ci.appveyor.com/project/madebr/x16-emulator/builds/28339440/job/p1usf8pegdwty00d/artifacts

manuvdp commented:
@madebr it works with your latest version!
Thx! Now learning from the sourcecode....
